### PR TITLE
Specified diagnostic for iterating known array type without --downlevelIteration

### LINKED ISSUE #16022: Given better error messages for iterated entities when not using --downlevelIteration
**TypeScript Version:**  2.3.3

**Code**

```ts
String.fromCharCode(...new Uint8Array(0));
```

**Expected behavior:**
I would expect there is no compile error.

**Actual behavior:**
image
image

I have to write this code:
```
String.fromCharCode(...Array.from(new Uint8Array(0)));
```

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Fixes #16022

If an entity can't be iterated for lack of a `[Symbol.iterator]` _and_ is one of a few known array-like types that would receive it in `--downlevelIteration`, a more specific error message indicating so is now emitted.

I think an ideal solution would have been to check whether the type _would_ be iterable in a higher output target, but I couldn't find prior art in code for doing so. Is there a better way to check for this case, or an existing or larger list of known names?

Refactors the big multi-level ternary previously used to generate the diagnostic message because it would have been even more multi-level from this change. Teeny improvement: `yieldType` isn't calculated unless it's needed now.

### CODE REVIEW

File: src/compiler/diagnosticMessages.json [hunk 1]
```diff
@@ -3016,7 +3016,10 @@
         "category": "Error",
         "code": 2792
     },
-
+    "Type '{0}' can only be iterated through when using the '--downlevelIteration' flag.": {
```
Comment (JoshuaKGoldberg) [inline src/compiler/diagnosticMessages.json:3019]: This is the error message suggested by @DanielRosenwasser in #16022. I'm wondering if it should also mention a higher output target?

```suggestion
    "Type '{0}' can only be iterated through when using the '--downlevelIteration' flag or with a '--target' of 'es2015' or higher.": {

```

Comment (DanielRosenwasser) [reply src/compiler/diagnosticMessages.json:3019]: That seems better, though what if a user can't do either? What's the alternative to `Array.from`? `Array.prototype.slice.call`?

Comment (JoshuaKGoldberg) [reply src/compiler/diagnosticMessages.json:3019]: 👍 or `[].slice.call` for short.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -33386,6 +33376,44 @@ namespace ts {
...
+            switch (n) {
+                case "Float32Array":
+                case "Float64Array":
+                case "Int16Array":
+                case "Int32Array":
+                case "Int8Array":
+                case "Uint16Array":
+                case "Uint32Array":
+                case "Uint8Array":
+                case "Uint8ClampedArray":
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:33413]: What's the one in the DOM that everyone hates? `NodeArray`?

Comment (DanielRosenwasser) [reply src/compiler/checker.ts:33413]: [`NodeList`](https://developer.mozilla.org/en-US/docs/Web/API/NodeList)

Comment (JoshuaKGoldberg) [reply src/compiler/checker.ts:33413]: 👌 classic.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -33354,18 +33354,8 @@ namespace ts {
...
-                            ? [Diagnostics.Type_0_is_not_an_array_type_or_does_not_have_a_Symbol_iterator_method_that_returns_an_iterator, true]
-                            : yieldType
-                                ? [Diagnostics.Type_0_is_not_an_array_type_or_a_string_type_Use_compiler_option_downlevelIteration_to_allow_iterating_of_iterators, false]
-                                : [Diagnostics.Type_0_is_not_an_array_type, true]
-                        : downlevelIteration
-                            ? [Diagnostics.Type_0_is_not_an_array_type_or_a_string_type_or_does_not_have_a_Symbol_iterator_method_that_returns_an_iterator, true]
-                            : yieldType
-                                ? [Diagnostics.Type_0_is_not_an_array_type_or_a_string_type_Use_compiler_option_downlevelIteration_to_allow_iterating_of_iterators, false]
-                                : [Diagnostics.Type_0_is_not_an_array_type_or_a_string_type, true];
+                    const disallowsStrings = !(use & IterationUse.AllowsStringInputFlag) || hasStringConstituent;
```
Comment (sandersn) [inline src/compiler/checker.ts:33357]: this should probably be `allowsStrings` too, and be inverted

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -33386,6 +33376,45 @@ namespace ts {
             }
 
             return arrayElementType;
+
+            function getIterationDiagnosticDetails(allowsStrings: boolean, downlevelIteration?: boolean): [DiagnosticMessage, boolean] {
```
Comment (sandersn) [inline src/compiler/checker.ts:33380]: downlevelIteration is always provided, so it doesn't need to be optional

**Review by sandersn — CHANGES_REQUESTED** [review]
Looks good save a couple of nitpicks. I think a static list of types is fine because I doubt the list will expand much over time.

**Comment by DanielRosenwasser** [comment]
@sandersn were you still going to merge this?

**Comment by sandersn** [comment]
Sorry, I forgot about this. I merged from master, and I'll merge the PR when tests finish running.

**Pull request merged**